active/0000-extensible-commands.md

```diff
@@ -0,0 +1,62 @@
+
+This lookup process would be refactored to use the DAG structure of loaded addons to build a flattened list of addons that implement the requested command. This stack would then be recursively invoked, such that each addon's version of the command can call `this._super()` to pass control to the next addon in the "stack". The bottom of the stack would be the built-in implementation (if one exists).
+
+This structure mimics Connect-style middleware, where each function in the stack can decide to pass control further, or halt the recursion by not calling `this._super()`.
+
+Here's a rough sketch in code:
+
+```
+var command, commandOptions;
+var orderedAddonList = project.addons;
+var commandStack = orderedAddonList.filter(function(addon) {
+        return addon.hasCommand(command);
+    }).map(function(addon) {
+        return addon.getCommand(command);
+    });
+
+var context = { _super: invokeNextCommandInStack };
+function invokeNextCommandInStack() {
+    var nextCommand = commandStack.shift();
+    return nextCommand.call(context, command, commandOptions);
+}
+
+invokeNextCommandInStack()
+```
+
+> **Note:** obviously, there's a lot more that should happen, and the commands are not functions themselves. Likely, either the Command model would need to incorporate the idea of a "stack", the `/lib/cli/cli.js` file would have to become stack-aware, or we introduce a new CommandStack abstraction to handle this.
+
+# Drawbacks
+
+* Extending built-in commands could result in user confusion, i.e. I checkout an Ember CLI project and run `ember test` and it performs unexpected behaviors, so I Google around for "ember test" and none of the standard docs help me.
+* Extending built-in commands could be a bit of a footgun - it would allow the addon developer to deviate from the conventions that this project attempts to cultivate.
+* Using the implementation approach outlined above, the addon that fires first for a built-in command could forgot to call super, causing a break in the chain and the default functionality not to run.
+
+# Alternatives
+
+* Rather than allowing complete overrides of commands, we could just provide hooks for common operations (i.e. modify the `availableOptions` hash, `anonymousOptions` array, etc). This would help prevent abuse, but limits use cases.
+* Even more restrictive, we could not allow any addon modification, and instead wait for common patterns or use cases to arise, and bake those into the built-in commands via optional arguments, auto-detecting addons, config, etc. This is difficult over the long term though, and doesn't scale well.
+* Finally, we may not want to allow customization of built-in commands at all, if the potential for abuse and surprising behavior outweighs the benefits of customization.
```

Comment:
I would disagree (no surprise there :wink:).

I view the CLI commands as the useful abstractions whose implementation details can vary. In the same way that I should be able to use `store.find('book')` without knowing exactly how the adapter fetches those books.

The intent is clear when I run `ember test`, and if I'm a new dev coming on to an existing Ember project, it makes it much simpler to onboard thanks to the conventional approach. I don't have to learn _this_ project's test command, generate command, etc, I just use the conventional commands, which may have been tweaked by the unique needs of this project, but still carry the same intent.

All that said, I understand the need to take away the magic and make things explicit. Perhaps there is a way it could be obvious when a built-in command is modified. Maybe the command output could show the stack as it's executing, so you get a sense of `super-test-addon:test` running, then built in `ember-cli:test`, etc.